src/main/java/de/zalando/aruha/nakadi/validation/EventValidation.java

```diff
@@ -1,39 +1,75 @@
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
 
-import java.util.Map;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
 
 public class EventValidation {
+    public static EventTypeValidator forType(final EventType eventType) {
+        return new EventTypeValidator(eventType);
+    }
 
-    private static final Map<String, EventTypeValidator> eventTypeValidators = Maps.newConcurrentMap();
+    public static JSONObject effectiveSchema(final EventType eventType) throws JSONException {
+        final JSONObject schema = new JSONObject(eventType.getSchema().getSchema());
 
-    public static EventTypeValidator forType(final EventType eventType) {
-        return forType(eventType, false);
+        switch (eventType.getCategory()) {
+            case BUSINESS: return addMetadata(schema);
+            case DATA: return wrapSchemaInData(schema);
+            default: return schema;
+        }
     }
 
-    public static EventTypeValidator forType(final EventType eventType, final boolean incrementalEdit) {
+    private static JSONObject wrapSchemaInData(final JSONObject schema) {
+        final JSONObject wrapper = new JSONObject();
 
-        final boolean contains = eventTypeValidators.containsKey(eventType.getName());
-        Preconditions.checkState(incrementalEdit || !contains, "Validator for EventType {} already defined",
-            eventType.getName());
+        normalizeSchema(wrapper);
 
-        if (!contains) {
-            final EventTypeValidator etsv = new EventTypeValidator(eventType);
-            eventTypeValidators.put(eventType.getName(), etsv);
-        }
+        wrapper.getJSONObject("properties").put("data_type", new JSONObject("{\"type\": \"string\"}"));
+
+        wrapper.getJSONObject("properties").put("data_op", new JSONObject("{\"type\": \"string\", \"enum\": [\"C\", \"U\", \"D\", \"S\"]}"));
+
+        wrapper.getJSONObject("properties").put("data", schema);
+
+        addToRequired(wrapper, new String[]{ "data_type", "data_op", "data" });
+
+        return wrapper;
+    }
 
-        return eventTypeValidators.get(eventType.getName());
+    private static JSONObject addMetadata(final JSONObject schema) {
+        normalizeSchema(schema);
+
+        schema.getJSONObject("properties").put("metadata", new JSONObject("{\"type\": \"object\"}"));
+
+        addToRequired(schema, new String[]{ "metadata" });
+
+        return schema;
     }
 
-    public static EventTypeValidator lookup(final EventType eventType) {
-        return eventTypeValidators.get(eventType.getName());
+    private static void addToRequired(final JSONObject schema, final String[] toBeRequired) {
+        final Set<String> required = new HashSet<>(Arrays.asList(toBeRequired));
+
+        for(int i = 0; i < schema.getJSONArray("required").length(); i++) {
+            required.add(schema.getJSONArray("required").getString(i));
+        }
+
+        schema.put("required", required);
```

Comment:
I did it this way to avoid duplicating entries. For example, if the user of this method asks to add "metadata" but "metatada" was already there, then it's better to it this way.